shiny-api/updater/updater.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/apiarian/migration-playground/shiny-api/common"
+)
+
+type Updater struct {
+	kc            *common.KafkaClient
+	command_topic string
+	done          chan struct{}
+	assignIDs     bool
+	nextID        int
+	mux           *sync.Mutex
+	entityCache   map[string]*common.Thing
+}
+
+func NewUpdater(
+	kc *common.KafkaClient,
+	command_topic string,
+	assignIDs bool,
+	nextID int,
+) *Updater {
+	return &Updater{
+		kc:            kc,
+		command_topic: command_topic,
+		done:          make(chan struct{}),
+		assignIDs:     assignIDs,
+		nextID:        nextID,
+		mux:           &sync.Mutex{},
+		entityCache:   make(map[string]*common.Thing),
+	}
+}
+
+func (u *Updater) Close() {
+	close(u.done)
+}
+
+func (u *Updater) DoYourThing() error {
+	return u.kc.RegisterMessageProcessor(
+		u.command_topic,
+		5*time.Minute,
+		u.done,
+		common.WrapThingCommandEntryProcessor(u.MakeCommandProcessor()),
+	)
+}
+
+func (u *Updater) MakeCommandProcessor() func(*common.ThingCommandEntry, error) {
+	return func(m *common.ThingCommandEntry, err error) {
+		if err != nil {
+			log.Printf("error before command could be processed: %v", err)
```

Comment:
Generally speaking, we've found that logging on a hotpath is an anti pattern in evented systems as you can get *lots* of events and see logs fill up.